recipes-graphics/graphene/graphene_%.bbappend

```diff
@@ -0,0 +1,4 @@
+# Disable ARM neon support. Backported from Arch Linux: https://github.com/archlinuxarm/PKGBUILDs/pull/1977
+# Upstream-status: Reported [https://github.com/ebassi/graphene/issues/215]
+EXTRA_OEMESON:append:arm = " -Darm_neon=false"
```

Comment:
Does `:arm` applies both to Aarch64 and ARMv7 or only to ARMv7?

If the crash only happens on ARMv7 i think it will make sense to only disable it there

Reply:
Just checked it. It only seems to apply to ARM-32 bits. Good

Reply:
But this still means we disable graphene NEON instructions unconditionally for everybody on ARMv7, regardless of whether webkitgtk is built or not. I think that's bad style.

Reply:
I understand your point but see a clear path to follow right now. It sounds like distros like Arch and ReHat just disabled NEON on graphene package for ARM: https://github.com/archlinuxarm/PKGBUILDs/pull/1977. If this is too much aggressive or not I don't have a strong enough argument to take position in one or another side.

Reply:
> But this still means we disable graphene NEON instructions unconditionally for everybody on ARMv7, regardless of whether webkitgtk is built or not. I think that's bad style.

The issue not only affects WebKitGTK. It also affects WPE when using the GTK4 plugin of Cog.

Graphene seems to be used by the gstreamer-gl plugin. It is needed to build the `gltransformation` and `glvideoflip` elements.

So removing NEON support on Graphene may impact performance  (would like to see some benchmarks) when using `gltransformation`,  `glvideoflip`  elements inside the gstreamer play pipeline.

Other than that, I'm not aware of any usage of Graphene when using WPE that may be impacted by this change.

Do you have any example of use case that will be impacted?

We have to make a difficult choice here, and I will rather have every ARMv7 user of WPE+Cog/GTK4 and WebKitGTK/GTK4 with a working build than a hypothetical performance improvement when playing videos with WPE/WebKitGTK with some non-standard configuration.

Ideally I would like to fix this without having to disable NEON support on Graphene. If you have any suggestion on how to do that it is welcome.

Reply:
> > But this still means we disable graphene NEON instructions unconditionally for everybody on ARMv7, regardless of whether webkitgtk is built or not. I think that's bad style.
> 
> The issue not only affects WebKitGTK. It also affects WPE when using the GTK4 plugin of Cog.

Right.

> Graphene seems to be used by the gstreamer-gl plugin. It is needed to build the `gltransformation` and `glvideoflip` elements.
> 
> So removing NEON support on Graphene may impact performance (would like to see some benchmarks) when using `gltransformation`, `glvideoflip` elements inside the gstreamer pipeline.

Correct.

> Other than that, I'm not aware of any usage of Graphene when using WPE that may be impacted by this change.
> 
> Do you have any example of use case that will be impacted?

My main use case is GStreamer as mentioned above. But note that this also impacts graphene use cases outside of meta-webkit.

> We have to make a difficult choice here, and I will rather have every ARMv7 user of WPE+Cog/GTK4 and WebKitGTK/GTK4 with a working build than a hypothetical performance improvement when playing videos with WPE/WebKitGTK with some non-standard configuration.
> 
> Ideally I would like to fix this without having to disable NEON support on Graphene. If you have any suggestion on how to do that, then please share it.

I know that there is no straight forward solution, but I think the neon packageconfig for graphene mentioned in https://github.com/Igalia/meta-webkit/pull/448#issuecomment-1424039235 is simple enough and won't surprise users of meta-webkit, since this needs to be done explicitly via local.conf.

Something like (untested):

**graphene.bbappend:**
PACKAGECONFIG[neon] = "-Darm_neon=true,-Darm_neon=false,"
PACKAGECONFIG += "neon"

**The build instructions could then mention:**
WebKitGTK and cog GTK4 plugin users should disable NEON instructions (see bug: FIXME)
PACKAGECONFIG:remove:arm = "neon"

Reply:
Ah, someone else posted the graphene approach already upstream: https://lore.kernel.org/openembedded-core/20230213120407.1092455-1-f_l_k@t-online.de/T/#u

Reply:
> Ah, someone else posted the graphene approach already upstream: https://lore.kernel.org/openembedded-core/20230213120407.1092455-1-f_l_k@t-online.de/T/#u

Good. If this gets merged in oe-core then we don't have to do anything here.
But the end result will be the same: NEON support on Graphene will get disabled by default for all ARMv7 users

Reply:
> **The build instructions could then mention:**
> On ARMv7 WebKitGTK and cog GTK4 plugin users should disable NEON instructions in their local.conf (see bug: FIXME)
> `PACKAGECONFIG:remove:pn-graphene = "neon"`

I agree with the PACKAGECONFIG option so it is possible to enable or disable it on local.conf

But I think things should work by default, so I would default it to be disabled and then add a note at https://github.com/Igalia/meta-webkit/wiki/PerformanceTips about enabling it for users that don't care about GTK

Reply:
> But OTOH that's not really meta-webkit exclusive.

Agreed. On top of that I'm a bit skeptical regarding whether enabling NEON or not support for Graphene has any noticeable performance improvement on WPE when playing videos with GStreamer. Have you tested it?

Reply:
Fixed on Yocto Poky mickledore: https://lists.openembedded.org/g/openembedded-devel/message/100996